portal/common/lib/page_fetching.ts

```diff
@@ -28,7 +29,19 @@ export async function resolveAndFetchPage(parsedUrl: DomainDetails): Promise<Res
     if (isObjectId) {
         console.log("Object ID: ", resolveObjectResult);
         console.log("Base36 version of the object ID: ", HEXtoBase36(resolveObjectResult));
-        return fetchPage(client, resolveObjectResult, parsedUrl.path);
+        // Rerouting based on the contents of the routes object,
+        // constructed using the ws-resource.json.
+        const routes = await getRoutes(client, resolveObjectResult);
+        if (!routes) {
+            console.warn("No routes found for the object ID");
+            return fetchPage(client, resolveObjectResult, parsedUrl.path);
+        }
+        let matchingRoute: string | undefined;
+        matchingRoute = matchPathToRoute(parsedUrl.path, routes)
+        if (!matchingRoute) {
+            console.warn(`No matching route found for ${parsedUrl.path}`);
+        }
+        return fetchPage(client, resolveObjectResult, matchingRoute ?? parsedUrl.path);
```

Comment:
Remark: seems to me here that we first try the routes, and then if there are no routes we try to get the object at the direct path. This may be problematic (imagine if we have an `index.html` with a `/*` route, but still have some other static assets like `/image.png`. we would never be able to load the image). 

Also, this is contrary to the _longest prefix matching_ that we do on the routes (i.e., the longest possible prefix is the full path, so we should match that first). 

Suggestion: if the above is correct, I would suggest to change to first trying the full path, and if that fails match the route. Maybe also start fetching the routes in the backgroud while the full path is fetched.

Reply:
Based on the [tests](https://github.com/MystenLabs/walrus-sites/blob/e86a2b6a5a76bf59ab1e37bad2c21635772569bf/portal/common/lib/routing.test.ts#L19-L35) this case is already covered and the longest prefix matching applies. 

i.e. given `/else.jpeg` the longest prefix will match first, i.e. both `/somewhere/else/*` and `/somewhere/else` will be checked first, and then since no matches are found in the specified routes, the last option is the fallback of `['/*', '/default.html']` which means _"catch everything else and redirect it to default"_. 

If `/else.jpeg` doesn't match the `/default.html`, it is a dev misconfiguration, isn't it?

Reply:
quick example to see if we are saying the same thing or not :)

let's say i have files:
- `/index.html`
- `/image.jpg`

and routes: `/* : /index.html`

i would expect this to work, and that all paths _except_ `/image.jpg`are routed to `/index.html`.

I.e., I would first check if there is a direct match, and if there isn't, check the routes.

Reply:
oh! I see, then yes this is not happening. I can make this work.

Reply:
but do you agree this should be the expected behavior? I'm not too sure what happens e.g., with vercel json

Reply:
It's a matter of taste. [Nextjs middleware matcher](https://nextjs.org/docs/app/building-your-application/routing/middleware) would match the pattern as I have implemented it initially.

The rule of  checking first for `/image.jpg` and if it's not found, then try to apply re-route to `/index.html` based on pattern matching is a rule that is implied, and I wouldn't say that it would be an expected behavior.

Meaning, if I have specified a pattern, to reroute to a specific path, then I would first expect for my rule to apply.

Reply:
For future us: After our call we concluded that it would be best to check first for a direct path match, and if no resource was fetched, then we can use the Routes object to resolve it. This would be compatible with SPAs.